src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteVisitor.cs

```diff
@@ -0,0 +1,96 @@
+
+namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
+{
+    internal abstract class CallSiteVisitor<TArgument, TResult>
+    {
+        protected virtual TResult VisitCallSite(IServiceCallSite callSite, TArgument argument)
+        {
+            var factoryService = callSite as FactoryService;
+            if (factoryService != null)
+            {
+                return VisitFactoryService(factoryService, argument);
+            }
+            var closedIEnumerableCallSite = callSite as ClosedIEnumerableCallSite;
+            if (closedIEnumerableCallSite != null)
+            {
+                return VisitClosedIEnumerable(closedIEnumerableCallSite, argument);
+            }
+            var constructorCallSite = callSite as ConstructorCallSite;
+            if (constructorCallSite != null)
+            {
+                return VisitConstructor(constructorCallSite, argument);
+            }
+            var transientCallSite = callSite as TransientCallSite;
+            if (transientCallSite != null)
+            {
+                return VisitTransient(transientCallSite, argument);
+            }
+            var singletonCallSite = callSite as SingletonCallSite;
+            if (singletonCallSite != null)
+            {
+                return VisitSingleton(singletonCallSite, argument);
+            }
+            var scopedCallSite = callSite as ScopedCallSite;
+            if (scopedCallSite != null)
+            {
+                return VisitScoped(scopedCallSite, argument);
+            }
+            var constantCallSite = callSite as ConstantCallSite;
+            if (constantCallSite != null)
+            {
+                return VisitConstant(constantCallSite, argument);
+            }
+            var createInstanceCallSite = callSite as CreateInstanceCallSite;
+            if (createInstanceCallSite != null)
+            {
+                return VisitCreateInstance(createInstanceCallSite, argument);
+            }
+            var instanceCallSite = callSite as InstanceService;
+            if (instanceCallSite != null)
+            {
+                return VisitInstanceService(instanceCallSite, argument);
+            }
+            var serviceProviderService = callSite as ServiceProviderService;
+            if (serviceProviderService != null)
+            {
+                return VisitServiceProviderService(serviceProviderService, argument);
+            }
+            var emptyIEnumerableCallSite = callSite as EmptyIEnumerableCallSite;
+            if (emptyIEnumerableCallSite != null)
+            {
+                return VisitEmptyIEnumerable(emptyIEnumerableCallSite, argument);
+            }
+            var serviceScopeService = callSite as ServiceScopeService;
+            if (serviceScopeService != null)
+            {
+                return VisitServiceScopeService(serviceScopeService, argument);
+            }
+            throw new NotSupportedException("Call site type is not supported");
```

Comment:
In the absence of proper pattern matching, let's change this to `$"Call sit type '{callSite.GetType()}' is not supported"` so we don't have to attach a debugger when we mess up.